Source/WebCore/Modules/model-element/HTMLModelElementEntity.cpp

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#if ENABLE(MODEL_ELEMENT) && ENABLE(MODEL_PROCESS)
+
+#include "config.h"
+#include "HTMLModelElementEntity.h"
+
+#include "DOMMatrix.h"
+#include "Logging.h"
+#include "ModelPlayer.h"
+#include <wtf/IsoMallocInlines.h>
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(HTMLModelElementEntity);
+
+Ref<HTMLModelElementEntity> HTMLModelElementEntity::create(const HTMLModelElement& modelElement)
+{
+    return adoptRef(*new HTMLModelElementEntity(modelElement));
+}
+
+HTMLModelElementEntity::HTMLModelElementEntity(const HTMLModelElement& modelElement)
+{
+    m_modelElement = WeakPtr { &modelElement };
+    RELEASE_LOG(ModelElement, "%p - Created HTMLModelElementEntity", this);
+}
+
+HTMLModelElementEntity::~HTMLModelElementEntity()
+{
+    RELEASE_LOG(ModelElement, "%p - Destroyed HTMLModelElementEntity", this);
+}
+
+RefPtr<ModelPlayer> HTMLModelElementEntity::modelPlayer() const
+{
+    if (!m_modelElement)
+        return nullptr;
+
+    return m_modelElement->modelPlayer();
+}
+
+const DOMMatrix& HTMLModelElementEntity::transform() const
+{
+    auto player = modelPlayer();
+    if (!player)
+        return DOMMatrix::create(TransformationMatrix::identity, DOMMatrixReadOnly::Is2D::No).get();
+
+    TransformationMatrix matrix = player->entityTransform();
+    return DOMMatrix::create(matrix, DOMMatrixReadOnly::Is2D::No).get();
+}
```

Comment:
This ends up creating a new object each time you invoke the getter, right? That violates: https://w3ctag.github.io/design-principles/#attributes-like-data. In particular we want

> `obj.attribute === obj.attribute`

to be true.

I suspect that instead something like `setView()` and `getView()` would be better. And perhaps they can be directly on the `model` element or do we have a strong reason for this intermediate object?

Reply:
Discussed offline, going to do away the whole HTMLModelElementEntity and flatten the get/setTransform into HTMLModelElement itself.